pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -28,54 +28,92 @@ import (
 
 func TestReaderPool_NewBinaryReader(t *testing.T) {
 	tests := map[string]struct {
-		lazyReaderEnabled     bool
-		lazyReaderIdleTimeout time.Duration
+		lazyReaderEnabled             bool
+		lazyReaderIdleTimeout         time.Duration
+		eagerLoadReaderEnabled        bool
+		persistLazyLoadedHeaderFn     func(blockId ulid.ULID) lazyLoadedHeadersSnapshot
+		expectedLoadCountMetricBefore int
+		expectedLoadCountMetricAfter  int
 	}{
 		"lazy reader is disabled": {
-			lazyReaderEnabled: false,
+			lazyReaderEnabled:            false,
+			expectedLoadCountMetricAfter: 0, // no lazy loading
 		},
 		"lazy reader is enabled but close on idle timeout is disabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: 0,
+			lazyReaderEnabled:            true,
+			lazyReaderIdleTimeout:        0,
+			expectedLoadCountMetricAfter: 1,
 		},
 		"lazy reader and close on idle timeout are both enabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: time.Minute,
+			lazyReaderEnabled:            true,
+			lazyReaderIdleTimeout:        time.Minute,
+			expectedLoadCountMetricAfter: 1,
+		},
+		"lazy reader preShutdownLoadedBlocks is present": {
+			lazyReaderEnabled:             true,
+			lazyReaderIdleTimeout:         time.Minute,
+			eagerLoadReaderEnabled:        true,
+			expectedLoadCountMetricBefore: 1, // the index header will be eagerly loaded before the operation
+			expectedLoadCountMetricAfter:  1,
+			persistLazyLoadedHeaderFn: func(blockId ulid.ULID) lazyLoadedHeadersSnapshot {
+				return lazyLoadedHeadersSnapshot{
+					IndexHeaderLastUsedTime: map[ulid.ULID]int64{blockId: time.Now().UnixMilli()},
+					UserID:                  "anonymous",
+				}
+			},
+		},
+		"no valid preShutdownLoadedBlocks is present": {
+			lazyReaderEnabled:             true,
+			lazyReaderIdleTimeout:         time.Minute,
+			eagerLoadReaderEnabled:        true,
+			expectedLoadCountMetricBefore: 0, // although eager loading is enabled, this test will not do eager loading because the block ID is not in the lazy loaded file.
+			expectedLoadCountMetricAfter:  1,
+			persistLazyLoadedHeaderFn: func(_ ulid.ULID) lazyLoadedHeadersSnapshot {
+				// let's create a random blockID to be stored in lazy loaded headers file
+				invalidBlockID, _ := ulid.New(ulid.Now(), rand.Reader)
+				// this snapshot will refer to invalid block, hence eager load wouldn't be executed
```

Comment:
Is the block invalid, or just not the one we'll try to load? If the latter, perhaps `otherBlockID` would be a clearer name.

Reply:
I renamed it to `fakeBlockID` because the blockID is not real. I also updated the method name from `persistLazyLoadedHeaderFn` to `createLazyLoadedHeadersSnapshotFn` because it is outdated.